staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go

```diff
@@ -126,17 +126,17 @@ func newAdmissionMetrics() *AdmissionMetrics {
 	// Admission metrics for a step of the admission flow. The entire admission flow is broken down into a series of steps
 	// Each step is identified by a distinct type label value.
 	step := newMetricSet("step",
-		[]string{"type", "operation", "rejected"},
+		[]string{"type", "operation", "rejected", "namespace"},
```

Comment:
how is a namespace label not a cardinality problem here? O(10,000) unique namespaces are envisioned in https://github.com/kubernetes/community/blob/master/sig-scalability/configs-and-limits/thresholds.md

Reply:
10k namespaces would be an exceptional case which we can remedy by leveraging https://github.com/kubernetes/enhancements/issues/2305. It's also an existing concern with request metrics which have far greater potential issues for cardinality (which we can also sidestep by explicitly setting bounds for label values).

Reply:
@s-urbaniak

Reply:
can you open an issue to track for 1.22? if this represents a regression for ~all clusters, it might need to be reverted before 1.22.0

Reply:
> can you open an issue to track for 1.22? if this represents a regression for ~all clusters, it might need to be reverted before 1.22.0

opened https://github.com/kubernetes/kubernetes/issues/104008

Reply:
I'd be surprised if it affected normal clusters. But reflecting on it, it actually does make sense that this would happen on a test cluster, since we tend to generate namespaces on a per-test basis. Even so, probably not a bad idea to revert.

Reply:
namespace churn can happen in normal clusters too... anything that uses namespaces for bounded operations (builds,  jobs, etc) is likely to frequently create/delete namespaces

Reply:
I also don't think it's a good idea for controllers to add namespace labels to metrics (e.g. https://github.com/kubernetes/kubernetes/pull/102523)

Reply:
we found the `apiserver_admission_controller_admission_duration_seconds_bucket` metric caused 1165536 time series on a cluster with 112 namespaces. without digging deeper this caused prometheus to consume ~7gb of ram within seconds rendering default nodes notready very fast.

Reply:
It's nice to have namespace visibility, though.. maybe we should have a mode where you can turn on a namespace metric explicitly via flags.

Reply:
I believe it's the churn, not the existing number of namespaces that's causing this @s-urbaniak

Reply:
agreed, yes, but the added cardinality amplifies.

Reply:
> It's nice to have namespace visibility, though..

I can agree with that point for counters and gauges, but I don't think this is safe or really needed for histograms since we will most likely never use this information to create SLOs or any latency-based alert.
In addition to that, the valuable information for the user here is to know which namespaces the rejection is coming from. This is a piece of information that could be present in the logs and we could narrow the metric responsibility to only report rejections and 5XX error codes that could then be used in actual alerts.